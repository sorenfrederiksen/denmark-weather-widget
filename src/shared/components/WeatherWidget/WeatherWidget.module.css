@value fonts "../../global-styles/fonts.css";
@value colors "../../global-styles/colors.css";
@value sansSerifBold, sansSerifRegular from fonts;
@value vejrGreen, logoRed, ligeNuRed, breakingYellow, sportBlue, lightAccentGrey, darkAccentGrey, textBlack from colors;

.weatherWidget {
  width: 100%;
  max-width: 320px;
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translateX(-50%) translateY(-50%);

  background: #fff;
  border: 1px solid ligeNuRed;
  box-shadow: 2px 2px 2px RGBA(0, 0, 0, 0.2);

  font-family: sansSerifRegular;
}

.weatherWidget__header {
  background: ligeNuRed;
  color: #fff;
  position: relative;
}

.weatherWidget__header__label {
  padding: 4px;
  background: #fff;
  color: ligeNuRed;
  position: absolute;
  top: 8px;
  left: 8px;
  font-family: sansSerifRegular;
  font-style: italic;
  font-size: 12px;
  text-transform: uppercase;
  letter-spacing: 2px;
}

.weatherWidget__header__search {
  font-family: sansSerifBold;
  padding-left: 74px;
}

.weatherWidget__body {
  position: relative;

  border-right: 1px solid lightAccentGrey;
  border-left: 1px solid lightAccentGrey;
  border-bottom: 1px solid lightAccentGrey;
}

.weatherWidget__body__icon {
  padding: 8px;
  position: absolute;
  left: 0;
  top: 0;
  width: 35%;
  max-height: 100%;

  svg {
    width: 90%;
    display: block;
    margin: 0 auto;
  }
}

.weatherWidget__body__list {
  list-style: none;
  padding: 0;
  margin: 0;
  padding-left: 35%;
}

.weatherWidget__body__list__item {
  list-style: none;
  padding: 8px;
  background: lightAccentGrey;
}

.weatherWidget__body__list__item:nth-child(2n) {
  background: #fff;
}

.weatherWidget__body__list__item__label {
  display: inline-block;
  font-family: sansSerifBold;
  margin-right: 8px;
  font-weight: 700;
}

.weatherWidget__body__list__item__weatherDirection {
  transition: transfrom 350ms 351ms ease;
  display: inline-block;
  width: 22px;
  height: 22px;
  position: absolute;
  right: 5px;
  bottom: 5px;
}
