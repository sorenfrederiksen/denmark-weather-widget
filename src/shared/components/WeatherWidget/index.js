// @flow
import React, { Component } from 'react';
import { get } from 'axios';
import styles from './WeatherWidget.module.css';
import WeatherImage from '../WeatherImage';
import Autocomplete from '../Autocomplete';
import cityList from '../../assets/data/cityList.json';
import { ReactComponent as WindDirectionIndicator } from '../../assets/images/WindDirectionIndicator.svg';

//
// Types
// --------------------------------------------------------------------

export type WeatherWidgetProps = {
  city: string,
  type: string,
  description: string,
  temp: number,
  tempMin: number,
  tempMax: number,
  humidity: number,
  windSpeed: number,
  windDirection: number,
};

type State = {
  cityList: Array<string>,
};

type BodyListItemProps = {
  label: string,
  value: string,
};

type WindSpeedIndicatorProps = {
  windSpeed: number,
  windowDirection?: number,
};

//
// Helpers
// --------------------------------------------------------------------

const BodyListItem = ({ label, value }: BodyListItemProps) => (
  <li className={styles.weatherWidget__body__list__item}>
    <span className={styles.weatherWidget__body__list__item__label}>{label}</span>
    <span className={styles.weatherWidget__body__list__item__value}>{value}</span>
  </li>
);

const WindSpeedIndicator = ({ windSpeed, windDirection }: WindSpeedIndicatorProps) => (
  <span>
    <span className={styles.weatherWidget__body__list__item__weatherSpeed}>{`${windSpeed} m/s`}</span>
    {windDirection !== undefined ? (
      <span
        className={styles.weatherWidget__body__list__item__weatherDirection}
        style={{
          transform: `rotate(${windDirection}deg)`,
        }}
      >
        <WindDirectionIndicator />
      </span>
    ) : null}
  </span>
);

const setUrlForCity = (city) => {
  if (typeof window !== 'undefined' && typeof document !== 'undefined') {
    const { title } = document;
    const { origin } = window.location;
    const newUrl = `${origin}?city=${city}`;
    window.history.replaceState({}, document.title, newUrl);
  }
};

//
// Component
// --------------------------------------------------------------------

class WeatherWidget extends Component<WeatherWidgetProps, State> {
  constructor(props: WeatherWidgetProps) {
    super(props);
    const { city, type, description, temp, tempMin, tempMax, humidity, windSpeed, windDirection } = props;

    this.state = { ...props, cityList: [] };
  }

  getCityList = async () => {
    try {
      const { data } = await get('http://localhost:8500/get-city-list');
      this.setState({ cityList: data });
    } catch (error) {
      throw error;
    }
  };

  getCityData = async (city) => {
    try {
      const { data } = await get(`http://localhost:8500/get-city-weather/${city}`);
      // I realise this looks ridiculous, and it is, but I've got to get out of here. Minor bug
      // to be resolved on the backend
      if (!Object.prototype.hasOwnProperty.call(data, 'windDirection')) {
        data.windDirection = undefined;
      }
      this.setState({ ...data });
    } catch (error) {
      console.log(error);
      throw error;
    }
  };

  onChangeCity = (newCity) => {
    this.getCityData(newCity);
    setUrlForCity(newCity);
  };

  componentDidMount() {
    this.getCityList();
  }

  render() {
    const {
      city,
      type,
      description,
      temp,
      tempMin,
      tempMax,
      humidity,
      windSpeed,
      windDirection,
      cityList,
    } = this.state;
    return (
      <div className={styles.weatherWidget}>
        <header className={styles.weatherWidget__header}>
          <div className={styles.weatherWidget__header__label}>Vejret</div>
          <div className={styles.weatherWidget__header__search}>
            <Autocomplete
              elementId="weather-widget-search"
              initialText={city}
              placeholderText="Enter a Danish city's name"
              options={cityList}
              onChangeValue={this.onChangeCity}
            />
          </div>
        </header>
        <div className={styles.weatherWidget__body}>
          <div className={styles.weatherWidget__body__icon}>
            <WeatherImage type={type} />
          </div>
          <ul className={styles.weatherWidget__body__list}>
            <BodyListItem label="Temperature" value={`${Math.round(temp)}Â°`} />
            <BodyListItem label="Humidity" value={`${humidity}%`} />
            <BodyListItem
              label="Wind"
              value={<WindSpeedIndicator windSpeed={windSpeed} windDirection={windDirection} />}
            />
          </ul>
        </div>
      </div>
    );
  }
}

export default WeatherWidget;
