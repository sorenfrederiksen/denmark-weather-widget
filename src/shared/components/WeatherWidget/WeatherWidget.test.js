import React from 'react';
import { shallow } from 'enzyme';
import WeatherWidget from './index';

describe('(Component) WeatherWidget', () => {
  let wrapper;

  const props = {};

  beforeEach(() => {
    wrapper = shallow(<WeatherWidget {...props} />);
  });

  it('should match snapshot.', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
