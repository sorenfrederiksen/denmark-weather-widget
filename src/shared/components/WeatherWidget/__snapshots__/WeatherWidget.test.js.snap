// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) WeatherWidget should match snapshot. 1`] = `ShallowWrapper {}`;
